### PR TITLE
Add a library to handle the retry logic.

### PR DESCRIPTION
**Pull request by calebbrown**
A following PR will use this library to handle retries for the GitHub API.

This library is written this way to facilitate testing, which I thought
was particularly important for retry related logic as it is hard to
cover all the various cases during manual testing.

### CODE REVIEW

File: internal/retry/request.go [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+type roundTripper struct {
+	inner http.RoundTripper
+	o     *Options
+}
+
+func (rt *roundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
+	var resp *http.Response
+	var err error
+	rr := NewRequest(r, rt.inner.RoundTrip, rt.o)
+	for !rr.Done() {
```
Comment (oliverchang) [inline internal/retry/request.go:244]: in what case would this be done already?

File: internal/retry/request.go [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+func NewRequest(r *http.Request, client func(*http.Request) (*http.Response, error), o *Options) *Request {
+	return &Request{
+		client: client,
+		r:      r,
+		o:      o,
+	}
+}
+
+func (r *Request) Done() bool {
+	return r.done || r.attempts >= r.o.maxRetries
```
Comment (oliverchang) [inline internal/retry/request.go:150]: does `attempts` include the first attempt? If so, semantically this should be `attempts > maxRetries` right?

File: internal/retry/request.go [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+	}
+
+	// Immediate success if the request is 2xx or 3xx
+	if http.StatusOK <= resp.StatusCode && resp.StatusCode < http.StatusBadRequest {
+		return r.onDone(resp, err)
+	}
+
+	if r.o.retryAfter != nil {
+		// Check if the Retry-After header is set
+		d := r.o.retryAfter(resp)
```
Comment (oliverchang) [inline internal/retry/request.go:195]: just curious: Why do we get the user to provide this function?

**Review by oliverchang — APPROVED** [review]

File: internal/retry/request.go (hunk 1 above)
Comment (calebbrown) [reply internal/retry/request.go:244]: Never, immediately after instantiation.

The forloop semantics are convenient for being able to make the retry library testable.

Technically you could implement this as:

```go
for {
  resp, err := rr.Do()
  if rr.Done() {
    break
  }
}
```

I'll add a comment making this clearer.

File: internal/retry/request.go (hunk 2 above)
Comment (calebbrown) [reply internal/retry/request.go:150]: Semantically yes. And I will change this to match.

This means `Options.maxRetries == 0` always results in a single attempt, but no retry.

File: internal/retry/request.go (hunk 3 above)
Comment (calebbrown) [reply internal/retry/request.go:195]: Mostly for logging reasons at the moment (the current location has logging support). I'll add a TODO there to move it here later when I'm more confident in its correctness.

File: internal/retry/request.go (hunk 1 above)
Comment (oliverchang) [reply internal/retry/request.go:244]: ack. I misread the code here actually. thanks!

**Pull request merged**